Bug 1891460: KubeletConfig: Allow only positive values for KubeReserved, SystemReserved, EvictionHard and EvictionSoft

harche opened the pull request:
Allow only positive values for KubeReserved, SystemReserved and EvictionHard

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1891460

Signed-off-by: Harshal Patil <harpatil@redhat.com>

<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->

**- What I did**
Validate if KubeReserved, SystemReserved and EvictionHard in KubeletConfig are positive or not. 

**- How to verify it**
https://bugzilla.redhat.com/show_bug.cgi?id=1891460#c0

**- Description for the changelog**
Allow only positive values for KubeReserved, SystemReserved, EvictionHard and EvictionSoft
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

harche commented:
/assign @runcom

sinnykumari commented on pkg/controller/kubelet-config/helpers.go:
```diff
@@ -170,8 +170,8 @@ func validateUserKubeletConfig(cfg *mcfgv1.KubeletConfig) error {
 				}
-				if q.Sign() == -1 {
-					return fmt.Errorf("KubeletConfiguration: %s reservation value cannot be negative in kubeReserved", rr.String())
+				if q.Sign() < 0 || q.IsZero() {
```
Do we want reserved value input > 0 ? 
Also see q.Sign() doc, return value of Sign() will be either -1, 0 or 1 https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go#L522

harche replied:
Yes, we want the reserved value input > 0. 

I tried to mirror how the kubelet validates these parameters, 

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/helpers.go#L221-L223

```golang
if quantity.Sign() < 0 || quantity.IsZero() {
	return nil, fmt.Errorf("eviction threshold %v must be positive: %s", signal, &quantity)
}
```

sinnykumari commented:
/approve

harche commented:
/test e2e-aws-serial

kikisdeliveryservice commented:
/assign @rphillips

kikisdeliveryservice commented:
Looking at the underlying BZ I also see: "value greater than memory capacity" causes error... so this wouldn't close the bz fully.

see:https://bugzilla.redhat.com/show_bug.cgi?id=1891460#c1

kikisdeliveryservice commented:
for now since this doesn't fully close the bz

/hold

harche commented:
/retest

harche commented:
/retest

harche commented:
/retest

harche commented:
/retest

harche commented:
/test e2e-aws-workers-rhel7

harche commented:
> Looking at the underlying BZ I also see: "value greater than memory capacity" and "non-digital" causes error... so this wouldn't close the bz fully.
> 
> see:https://bugzilla.redhat.com/show_bug.cgi?id=1891460#c1

We won't be supporting `Value greater than capacity` validation. I have closed a similar bug about that, https://bugzilla.redhat.com/show_bug.cgi?id=1881400. I have also highlighted that in the comment https://bugzilla.redhat.com/show_bug.cgi?id=1891460#c6. 

Other than that, I have added more test cases for 'non-digital' values and also added fixes the validation for `KubeReserved`, `SystemReserved`, `EvicationHard` and `EvictionSoft`. 

/cc @rphillips

rphillips commented:
/lgtm

kikisdeliveryservice commented:
> > Looking at the underlying BZ I also see: "value greater than memory capacity" and "non-digital" causes error... so this wouldn't close the bz fully.
> > see:https://bugzilla.redhat.com/show_bug.cgi?id=1891460#c1
> 
> We won't be supporting `Value greater than capacity` validation. I have closed a similar bug about that, https://bugzilla.redhat.com/show_bug.cgi?id=1881400. I have also highlighted that in the comment https://bugzilla.redhat.com/show_bug.cgi?id=1891460#c6.
> 
> Other than that, I have added more test cases for 'non-digital' values and also added fixes the validation for `KubeReserved`, `SystemReserved`, `EvicationHard` and `EvictionSoft`.
> 
> /cc @rphillips

Thanks for this @harche !!

/hold cancel
